sw/device/lib/base/status.h

```diff
@@ -49,42 +52,66 @@ extern "C" {
  * The module identifier value is interpreted as three 5-bit fields
  * representing the characters [0x40..0x5F] (e.g. [@ABC ... _]).
  */
-
-#define STATUS_FIELD_CODE ((bitfield_field32_t){.mask = 0x1f, .index = 0})
-#define STATUS_FIELD_ARG ((bitfield_field32_t){.mask = 0x7ff, .index = 5})
-#define STATUS_FIELD_MODULE_ID \
-  ((bitfield_field32_t){.mask = 0x7fff, .index = 16})
-#define STATUS_BIT_ERROR 31
 typedef struct status {
   int32_t value;
 } status_t;
 
-// Uses a GCC statement-expr to embed conrtol-flow inside an expression.
-#define TRY(expr_)            \
-  ({                          \
-    status_t status_ = expr_; \
-    if (status_.value < 0) {  \
-      return status_;         \
-    }                         \
-    status_.value;            \
+/**
+ * Converts a value into a status_t.
+ *
+ * This macro uses the GCC/clang `__builtin_types_compatible_p` extension to
+ * detect the type of the input expression and apply the appropriate
+ * conversion.  Once a more thorough refactoring of the DIFs is done, this
+ * can be eliminated.
+ *
+ * @param expr_ Either a `status_t`, `rom_error_t` or `dif_result_t`.
+ * @return The `status_t` representation of the input.
+ */
+#define INTO_STATUS(expr_)                                                \
+  ({                                                                      \
+    typeof(expr_) ex_ = (expr_);                                          \
+    status_t status_;                                                     \
+    if (__builtin_types_compatible_p(typeof(ex_), status_t)) {            \
+      memcpy(&status_, &ex_, sizeof(status_));                            \
+    } else if (__builtin_types_compatible_p(typeof(ex_), rom_error_t)) {  \
+      memcpy(&status_, &ex_, sizeof(status_));                            \
+    } else if (__builtin_types_compatible_p(typeof(ex_), dif_result_t)) { \
+      absl_status_t code;                                                 \
+      memcpy(&code, &ex_, sizeof(code));                                  \
+      status_ = status_create(code, status_module_id, __FILE__,           \
+                              code == kOk ? 0 : __LINE__);                \
+    } else {                                                              \
+      static_assert(                                                      \
+          __builtin_types_compatible_p(typeof(ex_), status_t) ||          \
+              __builtin_types_compatible_p(typeof(ex_), rom_error_t) ||   \
+              __builtin_types_compatible_p(typeof(ex_), dif_result_t),    \
+          "Expressions passed to INTO_STATUS() must be of type "          \
+          "`status_t`, `rom_error_t` or `dif_result_t`");                 \
+    }                                                                     \
+    status_;                                                              \
```

Comment:
I didn't know about `__builtin_types_compatible_p`, thanks!

nit: I think you included the `static_assert` in the `else` branch to emphasize that we don't need to handle that, but moving it to the top could be clearer:
```suggestion
#define INTO_STATUS(expr_)                                                \
  ({                                                                      \
      static_assert(                                                      \
      __builtin_types_compatible_p(typeof(ex_), status_t) ||          \
          __builtin_types_compatible_p(typeof(ex_), rom_error_t) ||   \
          __builtin_types_compatible_p(typeof(ex_), dif_result_t),    \
      "Expressions passed to INTO_STATUS() must be of type "          \
      "`status_t`, `rom_error_t` or `dif_result_t`");                 \
    typeof(expr_) ex_ = (expr_);                                          \
    status_t status_;                                                     \
    if (__builtin_types_compatible_p(typeof(ex_), status_t)) {            \
      memcpy(&status_, &ex_, sizeof(status_));                            \
    } else if (__builtin_types_compatible_p(typeof(ex_), rom_error_t)) {  \
      memcpy(&status_, &ex_, sizeof(status_));                            \
    } else if (__builtin_types_compatible_p(typeof(ex_), dif_result_t)) { \
      absl_status_t code;                                                 \
      memcpy(&code, &ex_, sizeof(code));                                  \
      status_ = status_create(code, status_module_id, __FILE__,           \
                              code == kOk ? 0 : __LINE__);                \
    }
    status_;                                                              \
```

Reply:
Done.